File: Data Structures & Algorithms/buy-and-sell-crypto/submission-0.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int left = 0, right = 1, profit = 0;

        while (left < prices.size() && right < prices.size()) {
            if (prices[right] < prices[left]) {
                left = right;
                right = left + 1;
            } else {
                profit = std::max(profit, prices[right] - prices[left]);
                right++;
            }
        }

        return profit;
    }
};

File: Data Structures & Algorithms/best-time-to-buy-and-sell-stock-ii/submission-0.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        dp = {}

        def _dfs(i, bought):
            if i == len(prices):
                return 0
            if (i, bought) in dp:
                return dp[(i, bought)]
            
            res = _dfs(i + 1, bought)

            if bought:
                res = max(res, prices[i] + _dfs(i + 1, not bought))
            else:
                res = max(res, -prices[i] + _dfs(i + 1, True))
            
            dp[(i, bought)] = res

            return res

        return _dfs(0, False)